pipeline/00-ingest.R

```diff
@@ -296,33 +296,92 @@ training_data_clean <- training_data_w_hie %>%
+  # and the maximum price of the most recent valid sale in the last n years,
+  # both EXCLUDING the current sale
   left_join(
     left_join(
       training_data %>%
         select(meta_pin, meta_sale_document_num, meta_sale_date),
       training_data %>%
         filter(!sv_is_outlier) %>%
-        select(meta_pin, meta_sale_date),
+        select(meta_pin, meta_sale_date, meta_sale_price) %>%
+        mutate(meta_sale_price = as.numeric(meta_sale_price)),
       by = "meta_pin",
       relationship = "many-to-many"
     ) %>%
-      # as.duration(1) excludes the same sale from being identified as within
-      # 3 years of itself
-      mutate(within_n_years = between(
-        meta_sale_date.x - meta_sale_date.y,
-        as.duration(1),
-        as.duration(years(params$input$n_years_prior))
-      )) %>%
-      # Distinct is necessary because of multicard sales
-      distinct() %>%
-      summarise(
+      # as.duration(1) excludes the target sale from being identified as within
+      # N years of itself
+      mutate(
+        within_n_years = between(
+          meta_sale_date.x - meta_sale_date.y,
+          as.duration(1),
+          as.duration(years(params$input$n_years_prior))
+        ),
+        price_within_n_years = ifelse(within_n_years, meta_sale_price, NA_real_)
+      ) %>%
+      # Distinct is necessary because of multi-card sales
+      distinct(
+        meta_pin,
+        meta_sale_date.x,
+        meta_sale_date.y,
+        .keep_all = TRUE
+      ) %>%
+      mutate(price_within_n_years = replace_na(price_within_n_years, 0)) %>%
+      summarize(
         meta_sale_count_past_n_years = as.numeric(
           sum(within_n_years, na.rm = TRUE)
         ),
+        meta_sale_price_past_n_years = max(price_within_n_years, na.rm = TRUE),
         .by = c("meta_pin", "meta_sale_document_num")
       ),
     by = c("meta_pin", "meta_sale_document_num")
   ) %>%
+  mutate(
+    # Construct the "strata price", which is the most recent available price for
+    # a given sale EXCLUDING that target sale's price. In descending order of
+    # preference, strata price is:
+    # - The max sale price of the past N years
+    # - The finalized Board of Review value of the prior year
+    # - The finalized certified value of the prior year
+    # - The finalized Board of Review value of the prior year
+    meta_sale_price_past_n_years = na_if(meta_sale_price_past_n_years, 0),
+    meta_strata_price = case_when(
+      !is.na(meta_sale_price_past_n_years) & meta_sale_price_past_n_years != 0 ~
+        meta_sale_price_past_n_years,
+      !is.na(meta_board_tot) & meta_board_tot != 0 ~
+        meta_board_tot * 10,
+      !is.na(meta_certified_tot) & meta_certified_tot != 0 ~
+        meta_certified_tot * 10,
+      TRUE ~ meta_1yr_pri_board_tot * 10
+    ),
+    # Construct price bins using grouped quantiles of the strata price. This is
+    # what actually gets used as a predictor in the model
+    meta_strata_1 = as.character(
```

Comment:
**[Question, non-blocking]** Why the `_1` suffix for this field?

Reply:
Just an artifact from the re-using fields first used by the condo model, which has two strata variables.